// auto-playwright/dist/sanitizeHtml.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sanitizeHtml = void 0;
exports.getSanitizeOptions = getSanitizeOptions;
const sanitizeHtmlLibrary = require("sanitize-html");
const DEFAULT_SANITIZE_TAGS = sanitizeHtmlLibrary.defaults.allowedTags.concat([
    "body",
    "button",
    "form",
    "img",
    "input",
    "select",
    "textarea",
    "option",
]);
const DEFAULT_SANITIZE_STYLES = undefined;
const DEFAULT_SANITIZE_CLASS_LIST = undefined;
function getSanitizeOptions() {
    return {
        // The default allowedTags list already includes _a lot_ of commonly used tags.
        // https://www.npmjs.com/package/sanitize-html#default-options
        //
        // I don't see a need for this to be configurable at the moment,
        // as it already covers all the layout tags, but we can revisit this if necessary.
        allowedTags: DEFAULT_SANITIZE_TAGS,
        // Setting allowedAttributes to false will allow all attributes.
        allowedAttributes: false,
        allowedClasses: DEFAULT_SANITIZE_CLASS_LIST,
        allowedStyles: DEFAULT_SANITIZE_STYLES,
    };
}
/**
 * The reason for sanitization is because OpenAI does not need all of the HTML tags
 * to know how to interpret the website, e.g. it will not make a difference to AI if
 * we include or exclude <script> tags as they do not impact the already rendered DOM.
 *
 * In my experience, reducing HTML only to basic tags produces faster and more reliable prompts.
 *
 * Note that the output of this function is designed to interpret only the HTML tags.
 * For instructions that rely on visual cues (e.g. "click red button") we intend to
 * combine HTML with screenshots in the future versions of this library.
 */
const sanitizeHtml = (subject) => {
    return sanitizeHtmlLibrary(subject, getSanitizeOptions());
};
exports.sanitizeHtml = sanitizeHtml;
